core/sdk-core/src/main/java/software/amazon/awssdk/core/endpointdiscovery/EndpointDiscoveryRefreshCache.java

```diff
@@ -87,6 +90,64 @@ public URI get(String accessKey, EndpointDiscoveryRequest request) {
         return endpoint.endpoint();
     }
 
+    public CompletableFuture<URI> getAsync(String accessKey, EndpointDiscoveryRequest request) {
+        key = accessKey;
+
+        // Support null (anonymous credentials) by mapping to empty-string. The backing cache does not support null.
+        if (key == null) {
+            key = "";
+        }
+
+        if (request.cacheKey().isPresent()) {
+            key = key + ":" + request.cacheKey().get();
+        }
+
+        EndpointDiscoveryEndpoint endpoint = cache.get(key);
+
+        if (endpoint == null) {
+            if (request.required()) {
+                return discoverEndpoint(request).handle(
```

Comment:
This method is pretty long and unwieldy. Can we break this branch out to a separate method?

Reply:
Yeah, suggesting the following:
```
if (endpoint != null) {
   return endpointFromCache(xxx);
}

if (request.isRequired()) {
  return discoverEndpoint(xxxx);
}

return temEndpoint(xxx);
```

Reply:
Sure, I will make this more readable

Reply:
I have modified this to reuse major chunks of the existing getSync method.